zebrad/tests/common/lightwalletd/sync.rs

```diff
@@ -0,0 +1,206 @@
+        while !is_zebrad_finished.load(Ordering::SeqCst) {
+            // Just keep checking the `lightwalletd` logs for errors.
+            // It usually logs something every 30-90 seconds,
+            // but there's no specific message we need to wait for here.
+            assert!(
+                lightwalletd_mut.wait_for_stdout_line(None),
+                "lightwalletd output unexpectedly finished early",
+            );
+        }
+
+        tracing::info!(?test_type, "waiting for lightwalletd to sync to the tip...");
+        while !are_zebrad_and_lightwalletd_tips_synced(lightwalletd_rpc_port, zebra_rpc_address)? {
+            let previous_check = Instant::now();
+
+            // To improve performance, only check the tips occasionally
+            while previous_check.elapsed() < TIP_CHECK_RATE_LIMIT {
+                assert!(
+                    lightwalletd_mut.wait_for_stdout_line(None),
+                    "lightwalletd output unexpectedly finished early",
+                );
+            }
+        }
+
+        // Tell the other thread that `lightwalletd` has finished
+        is_lightwalletd_finished.store(true, Ordering::SeqCst);
+
+        Ok(lightwalletd_mut)
+    };
+
+    // Run both threads in parallel, automatically propagating any panics to this thread.
+    std::thread::scope(|s| {
+        // Launch the sync-waiting threads
+        let zebrad_thread = s.spawn(|| {
+            zebrad_wait_fn().expect("test failed while waiting for zebrad to sync");
+        });
+
+        let lightwalletd_thread = s.spawn(|| {
+            let lightwalletd_result = lightwalletd_wait_fn();
+            is_lightwalletd_finished.store(true, Ordering::SeqCst);
+            lightwalletd_result.expect("test failed while waiting for lightwalletd to sync");
+        });
+
+        // Mark the sync-waiting threads as finished if they fail or panic.
+        // This tells the other thread that it can exit.
+        //
+        // TODO: use `panic::catch_unwind()` instead,
+        //       when `&mut zebra_test::command::TestChild<TempDir>` is unwind-safe
+        s.spawn(|| {
+            let zebrad_result = zebrad_thread.join();
+            is_zebrad_finished.store(true, Ordering::SeqCst);
+
+            zebrad_result.expect("test panicked or failed while waiting for zebrad to sync");
+        });
+        s.spawn(|| {
+            let lightwalletd_result = lightwalletd_thread.join();
+            is_lightwalletd_finished.store(true, Ordering::SeqCst);
+
+            lightwalletd_result
+                .expect("test panicked or failed while waiting for lightwalletd to sync");
+        });
+    });
+
+    Ok((lightwalletd, zebrad))
+}
+
+/// Returns `Ok(true)` if zebrad and lightwalletd are both at the same height.
+#[tracing::instrument]
+pub fn are_zebrad_and_lightwalletd_tips_synced(
+    lightwalletd_rpc_port: u16,
+    zebra_rpc_address: SocketAddr,
+) -> Result<bool> {
+    let rt = tokio::runtime::Builder::new_multi_thread()
```

Comment:
Does it need a multi-thread runtime? It probably doesn't matter either way since this is only being called once per minute, but I'm curious about the overhead for keeping it alive vs startup and shutdown.

Reply:
Yes, `tokio::block_in_place()` is used in the state, and it panics in a single-threaded runtime.